### ISSUE TITLE
Redisign the container navigation in browse mode to avoid loss of information for the user

### ISSUE THREAD

**Issue by Adriani90** [nvaccess/nvda#15732]
### Steps to reproduce:
1. With NVDA running, open this example in a browser:
```
<div>
<h1>hello</h1>
</div>
<main>
<h1>Article Title</h1>
<nav aria-label="Table of Contents">
<!-- Table of contents links -->
</nav>
<p>This is some article content.</p>
<h2>A Sub Heading</h2>
</main>
<footer>
<p>Copyright 2023 Example Corp</p>
</footer>
Note: this applies to every kind of nested containers such as tables, lists etc. The landmark is just an example.
2. Navigate the above example with d, shift+d, comma or shift+comma.

### Actual behavior:
* When using d or shift+d, NVDA skips the inner container which is nested into the main container.
* When using comma and shift+comma, NVDA skips the inner container. Additionally, NVDA does not navigate to the end of the current container, but to the beginning of the next container if there is one right afterwards.

### Expected behavior:
Scenario I:
1. Press d: you land on the main landmark
2. Press d again, you land on the nav landmark
3. press d again, you land on footer
4. Press d again: no further landmarks
Scenario II:
1. Press d: you land on the main landmark
2. Press Comma: virtual cursor moves to the last character in the container
3. Press comma again: NVDA reports "end of container"
4. Press shift+comma (virtual cursor moves to the first character in the container).
5. Press shift+comma again: NVDA reports "beginning of container
Scenario III:
1. Press d: you land on main landmark
2. Press d again: you land on nav landmark
3. Press comma or shift+comma: Same as scenario 2, steps 2-5 but the virtual cursor stays in the inner nav landmark.

### NVDA logs, crash dumps and other attachments:
n/a

### System configuration
#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
2023.3

#### Windows version:
Windows 11 22 H2

#### Name and version of other software in use when reproducing the issue:
All applications that support browse mode

#### Other information about your system:
n/a

### Other questions
#### Does the issue still occur after restarting your computer?
yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
yes, every version has this behavior.

#### If NVDA add-ons are disabled, is your problem still occurring?
yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
yes

**Comment by Emil-18** [comment]
The reason  it doesn't report the nav landmark is because it dont contain anything. When I added a <p> tag inside it, it worked fine

**Comment by seanbudd** [comment]
This seems like a proposed change of behaviour not a bug - we'd like more feedback before considering implementation

**Comment by Adriani90** [comment]
@seanbudd when looking at the user guide chapter 6.1 it says following:

---Name | ---Key | ---Description |
Move to start of container | shift+comma | Moves to the start of the container (list, table, etc.) where the caret is positioned |
Move past end of container | comma | Moves past the end of the container (list, table, etc.) where the caret is positioned |

So the current behavior does not match with the user guide, in my understanding this is a bug then.
moving to start of the current container means moving to the first character in that container.
Moving past the end of the container means moving out of the container but this does not make sense in practice. Why should an user expect to move past the end of a container? Actually the caret should move to the end of the container which is the last character.

cc: @XLTechie, @CyrilleB79  any thoughts?

**Comment by CyrilleB79** [comment]
@Adriani90 moving past the end of a container is instead very useful.

If I am in a table, a list or a section and if I want to ignore the remaining content of this container, I just have to press comma to go to the next information which is the first information that follows the table/list/section.

**Comment by Adriani90** [comment]
In the end it would mean people would have to press comma followed by down arrow to achieve the same result which is not a big pain. Especially for new users the current behavior is difficult to understand because using comma and shift+comma does not provide a reallly logical navigation patern. Von meinem iPhone gesendetAm 10.11.2023 um 01:01 schrieb Cyrille Bougot ***@***.***>:﻿
@Adriani90 moving past the end of a container is instead very useful.
If I am in a table, a list or a section and if I want to ignore the remaining content of this container, I just have to press comma to go to the next information which is the first information that follows the table/list/section.

**Comment by XLTechie** [comment]
@Adriani90 wrote:

> the current behavior is difficult to understand because using comma and shift+comma does not provide a reallly logical navigation patern.

Can you explain or elaborate this reasoning some more please? I think I somewhat 
agree with it, as in usual usage, move to end is not the same as move to next. 
That is, home and end: home takes you to start of line (shift+comma), and end 
takes you to end of line. If End was similar to comma, it would instead take you 
to the start of the next line.

That of course makes no sense. In my mind, comma in this usage has always been 
similar to End, but for a containing element. So going past the end of the 
element (to the next element), does seem a bit strange to me. But I thought 
perhaps I was just weird.

Also, I might be interested in harmonizing with what other screen readers do 
here, if there is a consistent paradigm.

**Comment by CyrilleB79** [comment]
I agree that other commands which deal with beginning and end of a specific part of the content (line, full document) jump to end, whereas this one jumps past the end, which is not consistent.

But IMO, jumping past the end of the container is much more useful and frequent than jumping to the last character of the container. If the container is a table, we can also record that there are commands to jumpt to the end of the row or column.
If the decision is made to change the current behaviour, I would appreciate an option to choose between old and new behaviour, or at least an easy interface to be able to restore the old behaviour through an add-on.

Regarding the other screen readers:
* Narrator does not seem to offer such container-related commands, so there is no possible comparison
* Jaws differs with NVDA in that it allows to put the BM (virtual) cursor on the "End of container" text, text which is part of the virtual buffer content; on the contrary in NVDA, "End of container" is not a text present in the virtual buffer but a notification heard when your cursor jumps from in the container to out of it, or, in braille, an appended text that cannot be jumped to with routing cursor. Thus when you press comma in Jaws, it litterally jumps on the end boundary of the container, not just before, nor after.

**Comment by Adriani90** [comment]
> IMO, jumping past the end of the container is much more useful and frequent than jumping to the last character of the container.

Do you have strong evidence that this is true? I hear from a lot of users, especially new users, that this is not intuitive at all and that's why they don't use this feature although it could be more useful than it is now. Moreover, when you have multiple containers and you press comma and shift+comma multiple times you jump in and out of containers without having a clue about what the logic of navigation is. More than that, the current behavior contradicts the user guide. Pressing shift+comma does not move always to the start of the container. Pressing comma does not move always past the end of the container, but often on the beginning of the next container.
Could you please give some concrete examples where jumping past the end of the container compared with jumping to the end of container and then pressing right arrow key or down arrow key brings a significant win of efficiency? I mean even when jumping out of the container you have to press a navigation key to explore the content after the container anyway.

> If the container is a table, we can also record that there are commands to jumpt to the end of the row or column.

This is not the same like jumping to the end of the container, imagine a table with 20 columns and 70 rows. I have to press more commands to reach the last cell.
* Either I press the command to the end of the row and then the command to the end of the column
* Or I press comma which brings me out of the table, but then I press left arrow to land back into the table which is contra intuitive in my opinion.

**Comment by CyrilleB79** [comment]
> > IMO, jumping past the end of the container is much more useful and frequent than jumping to the last character of the container.
> 
> Do you have strong evidence that this is true? I hear from a lot of users, especially new users, that this is not intuitive at all and that's why they don't use this feature although it could be more useful than it is now.

To clarify, I was only providing a feedback on my own usage, not a general picture on all users' usage.

> Moreover, when you have multiple containers and you press comma and shift+comma multiple times you jump in and out of containers without having a clue about what the logic of navigation is.

> More than that, the current behavior contradicts the user guide. Pressing shift+comma does not move always to the start of the container.

Could you provide an example?

> Pressing comma does not move always past the end of the container, but often on the beginning of the next container.

That's the case if there is nothing between the two containers. As an example, imagine two tables (enable layout tables):
`data:text/html,<table><tr><td>A</td><td>B</td></tr></table><table><tr><td>AA</td><td>BB</td></tr></table>`
When you are in the first table and you press comma, you will land on the first char in the second table because that's the first text appearing after having left the first table. Remember that unlike in Jaws, NVDA does not allow to put the virtual cursor on table boundary indications ("End of table" / "Table with x rows and y columns").


> Could you please give some concrete examples where jumping past the end of the container compared with jumping to the end of container and then pressing right arrow key or down arrow key brings a significant win of efficiency? I mean even when jumping out of the container you have to press a navigation key to explore the content after the container anyway.

Of course, but it makes just one less key to press.
On the opposite, if comma jumped on the last character of the container, in the case of a table, I feel very uncomfortable to explore the last cell starting from the last character: the point of interest of the last cell is its first character, not the last one.

> > If the container is a table, we can also record that there are commands to jumpt to the end of the row or column.
> 
> This is not the same like jumping to the end of the container, imagine a table with 20 columns and 70 rows. I have to press more commands to reach the last cell.
> 
>    * Either I press the command to the end of the row and then the command to the end of the column
> 
>     * Or I press comma which brings me out of the table, but then I press left arrow to land back into the table which is contra intuitive in my opinion.


Maybe you need frequently to jump to last cell and prefer to have a single keystroke rather than alt+control+End and then alt+control+pgDown,. In this case, targeting the last character of the last cell is not the good solution as explained above; I'd recommend to implement a "jump to last cell" command, as it exists in Jaws. This way, you can start reading the last cell from the beginning of it's text.

To summarize my position:
1. I am not completely opposed to this change if people find it useful and more intuitive. But would need to have the choice between new and old behaviour, at least through an add-on
2. If there is a need to jump to the last cell of a table, a dedicated table navigation command should be implemented instead of using the command to jump at the endof the container (or past it).
3. If the command to jump past the end of the container is counter-intuitive, before modifying it, we should ensure that it's not just the documentation explanation which is not clear. Specifically, the notion of container should be understood correctly, as well as the difference with Jaws experience regarding boundaries.

**Comment by Adriani90** [comment]
> More than that, the current behavior contradicts the user guide. Pressing shift+comma does not move always to the start of the container.

I have to correct my statement here, this at least seems to work as expected. But given the fact that comma brings you out of the container, shift+comma brings the user either at the beginning of the outer container and skips the inner container, or NVDA reports "not in a container" if there was no outer container. So the navigation is not intuitive because depending of what you disable in document formating settings, NVDA does not report if there is an outer of an inner container. So that's why for an user the navigation seems not logical. So I still vote for the virtual cursor to stay in the current container when pressing comma to avoid skipping inner containers and loss of information.

> On the opposite, if comma jumped on the last character of the container, in the case of a table, I feel very uncomfortable to explore the last cell starting from the last character

good point here. However for tables I am not saying we should not provide a key stroke especially for jumping to the last cell. My point is to reach the end of the table as soon as possible without leaving it.

> Of course, but it makes just one less key to press.

But then in contrary with the current behavior you have to press one key press more in order to reach the start of the container you just leaved.

> unlike in Jaws, NVDA does not allow to put the virtual cursor on table boundary indications ("End of table" / "Table with x rows and y columns").

That's why I proposed the last character, which could be also a white space or a linefeed or what ever is the last in the current container.

**Comment by CyrilleB79** [comment]
> > Of course, but it makes just one less key to press.
> 
> But then in contrary with the current behavior you have to press one key press more in order to reach the start of the container you just leaved.

Sorry, I do not understand. Can you fully describe the use case to which this last answer applies to?

**Comment by Adriani90** [comment]
> Sorry, I do not understand. Can you fully describe the use case to which this last answer applies to?

You can take the code piece from this issue description to test it on a minimal case. There is the main landmark and the inner nav landmark. When I put the virtual cursor in the inner nav landmark and press comma, I leave the inner landmark. When I press shift+comma, I don't reach the beginning of the inner landmark, but the beginning of the main landmark which spans around the inner landmark.

This makes navigation really inefficient in nested lists, nested tables and nested landmarks the the risk of loosing information for an user is high. So that's why I propose to navigate to the end of the current container with comma instead of leaving it, then you can press shift+comma to reach the beginning of the current container and not the beginning of another container that spans around it.

There are a lot of usecases for this. From concept boards to web based project management tools etc. They work alot with nested containers and the current behavior is not really efficient.

@seanbudd After more testing and carefully reading of the user guide I agree the correct definition for this change is rather a feature request than a bug. However, I hope you bear with me if I let this open as it is since the issue description is still clear enough.

**Comment by CyrilleB79** [comment]
> > Of course, but it makes just one less key to press.
> 
> But then in contrary with the current behavior you have to press one key press more in order to reach the start of the container you just leaved.

OK I get it: you are describing the case where you press comma and just after shift+comma.
Then my question is: if you go to the end of the container, why do you need just after to go to its start?

More generally, I am thinking to another situation where a container has a common boundary with a sub-container. As an example, take the following:
`# Start of example`
* Item A
* Item B:
  * Sub-item C
  * Sub-item D

And then a sentence out-of-the list
`# End of example`

With your proposal, if your cursor is on "Item A" and if you press comma, you will hear "list with 2 items  Item D" the cursor will be located on the character "D". This character is the last character of the main list but also the last character of the sub-list. Thus, if you press then shift+comma, you will be at the beginning of the sub-list, i.e. on the "I" of "Item C". I feel these steps a bit confusing.
What do you think of it? I agree that in this case of nested containers with a common boundary, there is probably no very clear solution.

**Comment by Adriani90** [comment]
> if you go to the end of the container, why do you need just after to go to its start?

Especially for skim reading reasons. Imagine there is a list with 120 items or a table with 200 rows, and I want to reach the lower part of the list or table faster, then I go to its end and navigate up. And then I might want to jump right at the start to reach some areas of the top of the container in a faster way. So usually I would like to play around with comma and shift+comma in the same container to skim read parts of it. Navigating out of it with comma just increases the risk of pressing shift+comma afterwards and landing on the start of the outer container which is not related.

This applies also for containers with a lot of text paragraphs or where list items contain alot of text.

> if you press then shift+comma, you will be at the beginning of the sub-list, i.e. on the "I" of "Item C". I feel these steps a bit confusing.

I agree, though in my opinion it would be less confusing than the current behavior skipping sub containers. At least Jaws can report if a container is nested and has sub containers and Jaws can report when you enter a nested list or someting like that. So if NVDA could report that as well, the confusion would be mittigated.

> What do you think of it? I agree that in this case of nested containers with a common boundary, there is probably no very clear solution.

Really good example, thanks for that. Actually with the current behavior on your example you would leave the container and its sub container, by pressing comma, and then you would press shift+l two times to reach the item a. Correct?
In my proposed behavior you would reach the end of the container and its sub container by pressing comma, and you can press shift+l only once to reach the item a.

**Comment by CyrilleB79** [comment]
All the examples you provide are realistic and I agree that the new behaviour you suggest allows to reduce the number of keypresses. All these examples are about reading text in a specific container, e.g. "I want to read something at the end of the container and then I want to read something at the beginning of the container". There are other situations however where the interesting text is just after the container; and in this case, the current behaviour uses less keypresses as already explained before.

All this discussion shows that there is no optimal behaviour for all situations. Hence my proposal to make the behaviour of comma configurable, i.e. jump at the end (last character) of the container or, as today, jump past the end of the container.
I would also be interested to hear other people (NV Access or community) for their opinion, especially to know which value may be suitable as a default.

**Comment by XLTechie** [comment]
If this is to become configurable, I think @Adriani90's solution is the most 
logical default, and probably the more common use cases..
It keeps you in the defined container space unless you intentionally leave it, 
instead of skittering around various parts of the nested container 
structure, in a "you can get there from here, but you can't 
get back the same way" fashion.

E.g. if I came to this feature knowing nothing, I would expect the behavior 
@Adriani90 desires.

One thing to consider, is how easily this could be described in the user guide, 
if made configurable. It would likely require an example of use for both 
situations, and therefore its own subchapter.

**Comment by CyrilleB79** [comment]
> It keeps you in the defined container space unless you intentionally leave it, instead of skittering around various parts of the nested container structure, in a "you can get there from here, but you can't get back the same way" fashion. E.g. if I came to this feature knowing nothing, I would expect the behavior @Adriani90 desires.

I understand that the goal of @Adriani90's proposal is to try to remain in the same container and to be able to turn back to the same location when pressing comma and then shift+comma (or the other way around).
The proposal is already an improvement to reach this goal.
However, keep in mind that it is not possible in all cases. More specifically, as described in the example of https://github.com/nvaccess/nvda/issues/15732#issuecomment-1807237131, in the case of nested containers when the inner sub-container is located at the end or at the beginning of the outer container.
